### PR TITLE
[WIP] Goals: Database entry for ideal monthly goal

### PR DESCRIPTION
**Pull request by shall0pass**
This is an attempt to expose some data from the goal templates to show a visual indication of progress.  **No UI changes are currently within this PR**.  I'm looking for help with that aspect.

Each time a goal template is run, a database value 'goal' within the zero_budgets table gets updated.  The intention is to use this value as the "100%" value that the budget cells can be compared against.

_Because the database has been modified, there is a migration within this PR.  I've not done this before, so the migration file was manually named.  If this is undesirable, I could use some additional instruction._

### CODE REVIEW

**Comment by youngcw** [comment]
For a while I've considered reorganizing the template processing to remove some of the duplication of checks.  This may be a good time to do that where you are starting to return multiple, similar values.

I had thought that it would be best that the processing for each template type only returned the ideal value, then the main priority/category loop would then check for over budgeting.  That would probably clean things up and make what each step needs/is doing more clear.

What you've done so far seems like a good step though.

**Comment by youngcw** [comment]
Does this ever remove a goal from the database if the template gets removed?

**Comment by youngcw** [comment]
Long term would a dedicated db table be needed for templates, or would another column in the zero_budgets table be enough to store template settings?

If we think a new table would be needed we should probably create that table here instead of modifying zero_budgets.

**Comment by shall0pass** [comment]
> Does this ever remove a goal from the database if the template gets removed? It may not be a problem if the existence of a template is checked before using the goal value.

Hmm.  Good point.  Currently no.  We could reset the value to 0 if all categories are being written from the monthly menu.  We wouldn't want to do it if a single month were run. I'll try to find time to update this.

> Long term would a dedicated db table be needed for templates, or would another column in the zero_budgets table be enough to store template settings?
>
> If we think a new table would be needed we should probably create that table here instead of modifying zero_budgets.

Are there any other fields we need?  I'm not really sure what the best schema would be.  Another text field that holds the #template line, but without '#template'?  A well thought out plan would be welcome.

**Comment by youngcw** [comment]
> Are there any other fields we need? I'm not really sure what the best schema would be. Another text field that holds the #template line, but without '#template'? A well thought out plan would be welcome.

I cant thing of any fields that would be needed other than the goal, and the template info. 

I figured either the template string, or a json string of the template like the parser creates would work.  I bet a json version would be easier to create from user gui inputs.  But I don't work with databases really at all, so I don't know what would be best or is do able.

File: packages/loot-core/migrations/1694438752000_add_goal_targets.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+BEGIN TRANSACTION;
+
+ALTER TABLE zero_budgets ADD column goal INTEGER DEFAULT 0;
+
+COMMIT;
```
Comment (MatissJanis) [inline packages/loot-core/migrations/1694438752000_add_goal_targets.sql:5]: **💭 thought:** ‏When doing DB migrations - please think three steps ahead: will we need to do further migrations? Will we want to rename or alter this column in the near future?

Adding a new column won't cause any problems. Neither will creating a new table.

But deleting/renaming/altering existing columns/tables will be either VERY hard or possibly even impossible due to how server-sync works.

**Review by MatissJanis — COMMENTED** [review]
Just one comment about the DB migration. Currently I'm not a user of the goal templates features. I'm happy to give you a rubber stamp approval (but preferably someone who uses them reviews the PR).

Will start doing more in-depth reviews when we start exposing this functionality in the UI :)

File: packages/loot-core/migrations/1694438752000_add_goal_targets.sql (hunk 1 above)
Comment (shall0pass) [reply packages/loot-core/migrations/1694438752000_add_goal_targets.sql:5]: I understand.  Unless someone steps up and is interested in developing the GUI, this PR is DOA in my mind.  There's no purpose to it if whichever DB migration introduced isn't utilized.  I don't have enough expertise to push this across the finish line alone and certainly don't want to destabilize syncing by moving in a direction that hasn't been fully developed.

**Comment by youngcw** [comment]
I've been looking into what it would take to change the color of the balance value to change based on the goal stored in the table. It's not the full gui, but it's a start. 

I was thinking to set the color green for met, orange for not met.  No goal would leave existing black and red colors.

**Comment by youngcw** [comment]
@MatissJanis I made some visual changes here to show if the a goal has been met or not. Thus needing a review by you.  The balance value will go green for a met goal, yellow for an unmet goal, then the same red, gray existing behavior.

One bug that is yet to be figured out is that when writing the goal value to the db the gui doesn't pick up on the change and will only get the value if the cache is refreshed.

**Comment by MatissJanis** [comment]
As long as the feature is under a feature flag (which I think it still is): I'm happy with whatever is proposed here. I trust you folks to work on this independently. And @shall0pass has merge rights, so once it's approved - it's good to go. ;)

But I had an idea.. would it be possible to do all of the UI implementation by still using the `notes` field as the "database"? IMO that would be a safer approach: first build the UI around the existing `notes` field. And once we are happy with the feature and are confident in the data structure - then we create the real database table.

**Comment by youngcw** [comment]
> would it be possible to do all of the UI implementation by still using the `notes` field as the "database"?

I that definitely is an option.  I guess that would also make any needed changes not be set in stone as much.  Ill look at it.

**Pull request closed without merging**